### DISCUSSION TITLE
Fibaro RGBW2 Messages

### DISCUSSION THREAD

**Discussion by amilanov75** [zwave-js/zwave-js#4204] [General] [answered]
I have started testing the Fibaro RGBW2 module using the nr-zwave-js node-red-contrib and would greatly appreciate an explanation from someone that understands the back-end, as to why I am seeing 6 or 7 messages via a zniffer when sending a command to a Fibaro RGBW2 module to change colour.

In node-red I can see the following messages, and for the time being am assuming they are aligned somehow to the zniffer messages, however stand to be corrected:

a) If I send a hexcolor payload with property "hexcolor", I get 7x messages back immediately and 1x deferred:
-- 3x messages, one each with R, B & G value
-- 2x messages, each with R, G, B and W values in the single message (one reporting currentColor and the other targetColor)
-- 1x message with the hexcolor
-- 1x delayed message (about 4-5 seconds later) with the power usage

b) If I send payload with R, G, B and W specified and property "targetColor", I get 7x message back immediately and 1x deferred:
-- 4x messages, one each with R, B, G and WW values
-- 1x message with R, G, B and W values in the single message 
-- 1x message with the hexcolor
-- 1x delayed message (about 4-5 seconds later) with the power usage

So much of the above is wasted/duplicated, however in NR it could be that a lot of the data is just read from the DB in which case, great, however the number of messages I can see via a zniffer are real. 

I have the device added in S2 (and it's a hassle to pull apart to exclude and include again with no security), so can't readily see the detail of what's being sent via zniffer (but if that's what required to get to the bottom of this, just tell me and I'll go and re-include without security, and retest), so stand to be corrected that it's other data content that is being sent, however, either way 6-7 messages is a huge overload on the network if you have multiple physical modules to change the colour of at once - as I do, and when really you only need 1 message (or 2 depending on how you structure it) to communicate the colour of a 4x channel (RGBW) device.

**Comment by AlCalzone** [comment, marked as answer]
I think this can be explained by the fact that Z-Wave's `Color Switch CC` has a different set of commands than required by the API which most applications expect. Especially `hexColor` is just a convenience wrapper around the individual R, G and B channels. And the compound `currentColor` and `targetColor` are convenience wrappers for the individual channels' `currentColor` and `targetColor`.

> If I send a hexcolor payload with property "hexcolor"

This sends a single command to the device with the R, G and B channels. When that succeeds, the individual channels are optimistically updated to match the hex color. ( _3x messages, one each with R, B & G value_ )
The device then reports its current state, resulting in an update for the current and target value, as reported by the device. ( _2x messages, each with R, G, B and W values in the single message_ ). The hexColor state is synchronized with that ( _1x message with the hexcolor_ ).
The delayed power report is emitted when the device report it.


> If I send payload with R, G, B and W specified and property "targetColor"

This also sends a single command to the device with the R, G, B and WW channels. When that succeeds, the individual currentColor are optimistically updated ( _4x messages, one each with R, B, G and WW_ )
The device then reports its current state, resulting in an update for the current value, as reported by the device. The targetValue was set by you previously, so there is no update for that. ( _1x message with R, G, B and W values_ )
Also the hexColor is synchronized with the individual channels ( _1x message with the hexcolor_ ).
And the delayed power report is emitted when the device sends it.

So unless the NR plugin does something special, each of these should result in 3 commands total being sent: Color SET, Color Switch REPORT and later Power REPORT. Note that value update events don't necessarily translate to Z-Wave commands 1:1.
The color change is only actively verified when setting one individual color channel, otherwise the amount of traffic would be too high. 

FYI: You can load your network keys into Zniffer in order to decrypt the traffic.

**Comment by amilanov75** [comment]
Hey @AlCalzone 

Thank you for your prompt response. 

**Commands vs Messages**
I get from when you say "3 commands total being sent" that doesn't relate on a 1:1 basis to the messages being received (a little lesson for me there), as such it's good to hear your confirmation of what I am seeing as the actual messages/traffic via the zniffer.

**Optimistic Messages**
It is also very helpful to hear that the first batch of messages are the optimistic ones, as I believe there is the option to stop these from happening at the driver level, or at least I think there is. I will be reaching out to Marcus to see what he has done in the nr-wave-js node and also doing some further testing.

**Decrypting Traffic**
As for this comment "You can load your network keys into Zniffer in order to decrypt the traffic."
I give up. I tried copying and pasting every key in, but none worked.

**Odd results**
In the end I excluded the RGBW2 so I could try to see what was happening and saw some weird results. Less messages being sent for colour updates and more for brightness when added with no security vs when added with S2 (no key). I guess this is expected weirdness?

**Reply by AlCalzone** [reply]
> I guess this is expected weirdness?

I would have to see the actual messages to give my assessment on this.

---

There is a little bit of a trick for the security keys. Make a file `<HOMEID>.txt` where you replace `<HOMEID>` with your home ID as hex, in uppercase. In that file, enter
```
98;0102030405060708090A0B0C0D0E0F10;1
9F;aabbccddeeff00112233445566778899;1
9F;aabbccddeeff00112233445566778899;1
9F;aabbccddeeff00112233445566778899;1
```
where the things between the semicolons are your keys (S0 behind 98, S2 keys behind 9F). Then you can press this button to load them all:
![grafik](https://user-images.githubusercontent.com/17641229/153285438-df9662f9-3d4c-417f-bb01-7757f9a9122b.png)

The only caveat is that you have to capture the initial nonce exchange for Zniffer to be able to decode it.

**Comment by marcus-j-davies** [comment]
Optimistic Value Updates can be disabled here.

![image](https://user-images.githubusercontent.com/55892693/153291881-dcf5d8ec-6e71-405d-bca2-df3217ca2555.png)

which sets the driver option to enabled (as odd as it sounds 😃 )
https://github.com/zwave-js/node-red-contrib-zwave-js/blob/ea6aab997fe192ffed390b68c556c7a5b05d0ccc/zwave-js/zwave-js.js#L238

**Comment by amilanov75** [comment]
@marcus-j-davies 
I'll post some results at a later date, but I was still seeing the 3x optimistic results with the Fibaro RGBW2 module when that checkbox was selected.

**Reply by marcus-j-davies** [reply]
Unless you're not applying the controller changes - that will be the driver - and will need a bug report.
I get the expected behaviour - but then I don't use the Color Switch

**Reply by AlCalzone** [reply]
@amilanov75 keep in mind that when setting multiple colors, the individual channels' value still need to be synchronized. And since we don't poll after a multi-color SET, this is necessary for devices that don't call in the change on their own.

**Comment by amilanov75** [comment]
Thanks @AlCalzone

I wrote an email to you, can you check it please.
